calyx/src/passes/infer_share.rs

```diff
@@ -0,0 +1,386 @@
+        if comp.cells.iter().any(|cell| {
+            let cell_ref = cell.borrow();
+            cell_ref.is_reference()
+                || match cell_ref.type_name() {
+                    Some(name) => self.no_share.contains(name),
+                    None => false,
+                }
+        }) {
+            self.no_share.insert(comp.name.clone());
+            return Ok(Action::Continue);
+        }
+
+        //build the domination map
+        let mut dmap = DominatorMap::new(
+            &mut comp.control.borrow_mut(),
+            comp.name.id.clone(),
+        );
+        if self.print_dmap {
+            println!("{dmap:?}");
+        }
+
+        for (node, dominators) in dmap.map.iter_mut() {
+            //get the reads
+            let mut reads: ReadSet = ReadSet::default();
+            if let Some(c) =
+                DominatorMap::get_control(*node, &comp.control.borrow())
+            {
+                reads.get_reads_from_control(
+                    c,
+                    &self.state_shareable,
+                    is_begin_id(c, *node),
+                );
+            }
+
+            //if read and write occur in same group/invoke, then we cannot label it
+            //shareable. So we remove node from its dominators
+            dominators.remove(node);
+            for cell_name in reads.reads.clone() {
+                let key = NameSearch::new(cell_name);
+                if !key.is_written(dominators, comp, &self.state_shareable) {
+                    self.no_share.insert(comp.name.clone());
+                    return Ok(Action::Continue);
+                }
+            }
+        }
+        comp.attributes.insert("state_share", 1);
+        self.state_shareable.add(comp.name.clone());
+        Ok(Action::Continue)
+    }
+}
+
+///Contains the ids of all the cells that are read from in a given "node" in
+///the domination map graph
+#[derive(Default)]
+struct ReadSet {
+    pub reads: HashSet<ir::Id>,
+}
+impl ReadSet {
+    //given a port, insert the port's parent's id if the parent the port's parent
+    //is shareable
+    fn add_parent_if_shareable(
+        &mut self,
+        share: &ShareSet,
+        port: &ir::RRC<ir::Port>,
+    ) {
+        if let ir::PortParent::Cell(cell) = &port.borrow().parent {
+            if share.is_shareable_component(&cell.upgrade()) {
+                self.reads.insert(cell.upgrade().borrow().name().clone());
+            }
+        }
+    }
+
+    //if the assignment reads only dones, return true. This is used so that we
+    //can ignore reads of "done" cells.
+    fn reads_only_dones(assignment: &ir::Assignment) -> bool {
+        Self::is_done_port(&assignment.src)
+            && Self::guard_only_dones(&assignment.guard)
+    }
+
+    //returns true if port is a "done" port or is a constant
+    fn is_done_port(port: &ir::RRC<ir::Port>) -> bool {
+        port.borrow().name.id == "done" || port.borrow().is_constant(1, 1)
```

Comment:
This doesn't seem to be checking if something is a done port. Things that are not named `done` can be done ports if they have the `@done` attribute. Furthermore, the `is_constant` call not checking whether something is a done call at all